src/common/models.py

```diff
@@ -0,0 +1,137 @@
+from __future__ import annotations
+
+import datetime as dt
+import json
+from dataclasses import asdict, dataclass
+from enum import Enum, auto, unique
+from typing import TYPE_CHECKING
+
+if TYPE_CHECKING:
+    from mypy_boto3_batch.type_defs import KeyValuePairTypeDef
+
+
+@unique
+class ProcessingOutcome(Enum):
+    """Potential outcomes for granule processing
+
+    Individual jobs may succeed or fail in various ways, but ultimately
+    the processing of a granule may only be a success or failure.
+    """
+
+    SUCCESS = auto()
+    FAILURE = auto()
+
+
+@unique
+class JobOutcome(Enum):
+    """Potential outcomes for a compute job (e.g., with AWS Batch)
+
+    This enum describes a job attempt of processing a granule. Failures of a
+    job may be retryable due to infrastructure or ephemeral issues.
+    """
+
+    SUCCESS = auto(), ProcessingOutcome.SUCCESS
+    FAILURE_RETRYABLE = auto(), ProcessingOutcome.FAILURE
+    FAILURE_NONRETRYABLE = auto(), ProcessingOutcome.FAILURE
+
+    @property
+    def processing_outcome(self) -> ProcessingOutcome:
+        """The ProcessingOutcome for this job outcome"""
+        return self.value[1]
+
+
+HLS_GRANULE_ID_STRFTIME = "%Y%jT%H%M%S"
+
+
+@dataclass
+class GranuleId:
+    """Granule identifier"""
+
+    product: str  # Should be "HLS"
+    platform: str  # Should be one of ["L30", "S30"]
+    tile: str
+    begin_datetime: dt.datetime
+    version: str  # should be "v2.0"
+
+    @classmethod
+    def from_str(cls, granule_id: str) -> GranuleId:
+        """Parse components from a string ID"""
+        product, platform, tile, begin_datetime, version_major, version_minor = (
+            granule_id.split(".")
+        )
+        return cls(
+            product=product,
+            platform=platform,
+            tile=tile,
+            begin_datetime=dt.datetime.strptime(
+                begin_datetime, HLS_GRANULE_ID_STRFTIME
+            ),
+            version=".".join([version_major, version_minor]),
+        )
+
+    def to_str(self) -> str:
```

Comment:
Consider renaming `to_str` to `__str__` and changing expressions elsewhere from `granule_id.to_str()` to `str(granule_id)`. This would also allow simpler usage in f-strings (`f"...{granule_id}..."` rather than `f"...{granule_id.to_str()}..."`).

Reply:
Nice!